pkg/notification/notice.go

```diff
@@ -0,0 +1,170 @@
+
+	if v.skipUpdateCheck && v.telemetryDisabled {
+		logger.Debug("Skipping update check and metric ping")
+		return
+	}
+
+	go func() {
+		args = getFlags(args)
+		client := &http.Client{
+			Timeout: 3 * time.Second,
+		}
+
+		req, err := http.NewRequestWithContext(ctx, http.MethodGet, v.updatesApi, http.NoBody)
+		if err != nil {
+			logger.Warn("Failed to create a request for Trivy api", log.Err(err))
+			return
+		}
+
+		// if the user hasn't disabled metrics, send the anonymous information as headers
+		if !v.telemetryDisabled {
+			req.Header.Set("Trivy-Identifier", uniqueIdentifier())
+			req.Header.Set("Trivy-Command", strings.Join(args, " "))
+			req.Header.Set("Trivy-OS", runtime.GOOS)
+			req.Header.Set("Trivy-Arch", runtime.GOARCH)
+		}
+
+		req.Header.Set("User-Agent", fmt.Sprintf("trivy/%s", v.currentVersion))
+		resp, err := client.Do(req)
+		if err != nil || resp.StatusCode != http.StatusOK {
+			logger.Debug("Failed getting response from Trivy api", log.Err(err))
+			return
+		}
+
+		defer resp.Body.Close()
+		if err := json.NewDecoder(resp.Body).Decode(&v.latestVersion); err != nil {
+			logger.Debug("Failed to decode the Trivy response", log.Err(err))
+			return
+		}
+
+		// enable priting if update allowed and quiet mode is not set
+		if !v.skipUpdateCheck && !v.quiet {
+			v.responseReceived = true
+		}
+		v.done = true
+	}()
+}
+
+// PrintNotices prints any announcements or warnings
+// to the output writer, most likely stderr
+func (v *VersionChecker) PrintNotices(output io.Writer) {
+	if !v.responseReceived {
+		return
+	}
+
+	var notices []string
+
+	notices = append(notices, v.latestVersion.Warnings...)
+	for _, announcement := range v.latestVersion.Announcements {
+		if time.Now().Before(announcement.ToDate) && time.Now().After(announcement.FromDate) {
+			notices = append(notices, announcement.Announcement)
+		}
+	}
+
+	if v.currentVersion != v.latestVersion.Trivy.LatestVersion {
+		notices = append(notices, fmt.Sprintf("Version %s of Trivy is now available, current version is %s", v.latestVersion.Trivy.LatestVersion, v.currentVersion))
+	}
+
+	if len(notices) > 0 {
+		fmt.Fprintf(output, "\n📣 \x1b[34mNotices:\x1b[0m\n")
+		for _, notice := range notices {
+			fmt.Fprintf(output, "  - %s\n", notice)
+		}
+		fmt.Fprintln(output)
+		fmt.Fprintln(output, "To suppress version checks, run Trivy scans with the --skip-version-check flag")
+		fmt.Fprintln(output)
+	}
+}
+
+func (v *VersionChecker) LatestVersion() string {
+	if v.responseReceived {
+		return v.latestVersion.Trivy.LatestVersion
+	}
+	return ""
+}
+
+func (v *VersionChecker) Announcements() []announcement {
+	if v.responseReceived {
+		return v.latestVersion.Announcements
+	}
+	return nil
+}
+
+// getFlags returns the just the flag portion without the values
+func getFlags(args []string) []string {
+	var flags []string
+	for _, arg := range args {
+		if strings.HasPrefix(arg, "-") {
+			flags = append(flags, strings.Split(arg, "=")[0])
```

Comment:
perhaps we want to convert Shorthand of flags here.
I mean `-d` => `--debug`

Reply:
I'm not sure I understand - you mean that if it's a `-d` it should be appednded as `debug` - if so, can do!

Reply:
IIUC you will collect information about the flags used.

But some flags support short name (`d` - `debug`, `q` - `quiet`, etc.).
For proper statistics we should concatenate the short name and the full name. We can do this on the client or on the server.

I suggest doing it on the client

Reply:
I'm not sure about this - its going to be quite intrusive to have to iterate over all the flag groups and their flags to find the corresponding name for the short name.

Also, there is a good argument for not conflating the shortname with the long name because it would hide the information about what flags are being used. It might actually be useful to know that 47% of people use `--debug` rather than `-d`

Reply:
No problem, I just saw it and wrote my thoughts.
Do as you see fit 👍